shared/state.rs

```diff
@@ -43,12 +43,18 @@ impl State {
     }
 
     pub fn get_next_task_id(&mut self) -> Option<usize> {
-        for (id, task) in self.tasks.iter() {
-            if task.status == TaskStatus::Queued {
-                return Some(*id);
-            }
-        }
-        None
+        return self
+            .tasks
+            .iter()
+            .filter(|(_, task)| task.status == TaskStatus::Queued)
+            .filter(|(_, task)| {
+                task.dependencies
+                    .iter()
+                    .flat_map(|id| self.tasks.get(id))
+                    .all(|task| task.status == TaskStatus::Done)
```

Comment:
I feel like we need to handle the case that tasks fail. (`Killed`, `Failed`).
Right now this would result in tasks being stuck in `Queued` forever without obvious reason.

1. Either show the tasks the affected tasks is depending on in the `pueue status` overview. (of course this column should only be displayed, if there is any active dependency in the current state)
2. Automatically fail tasks that depend on other failed tasks. This should also result in some kind of error message output that can be checked via `log`.

Reply:
After thinking a little bit about this, automatic failure of all `children` seems to be the best way.

As a consequence, we should also think about a `--dependants/-d` flag to `restart`. If this flag is set, all tasks that depend on the task to restart, will be restarted as well.

Reply:
```
        5
       / \
      3   4
     / \
    1   2
```

Example:
Task 1 and 4 fail.
Task 2 succeeds.
Task 3 and 5 fail automatically.

Now do a `pueue restart 1 -d`:
```
        9
       / \
      7   8?
     /
    6
```


- Restart task 1 with id 6
- Restart task 3 as task 7, depending on task 6
- Restart task 6 as task 9, depending on task 7
- Should task 4 be restarted as well? It's part of the dependency graph, but it's also a leaf (and thereby a starting point) of the graph.

Reply:
It seems to be interesting, but we may have some weird case like :
```
  4
 / \
2   3
 \ /
  1
```
In this case task 1 need to be restarted only once.

I will have a look at this feature, maybe we should create another pull request / issue